### PR TITLE
chore(SetTheory/Ordinal/Notation): use `NatCast`

### PR DESCRIPTION
**Pull request by vihdzp**
... instead of using a custom `ofNat` function.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/SetTheory/Ordinal/Notation.lean [hunk 1]
```diff
@@ -87,11 +89,7 @@ def repr' (prec : ℕ) : ONote → Format
       ("oadd " ++ (repr' max_prec e) ++ " " ++ Nat.repr (n : ℕ) ++ " " ++ (repr' max_prec a))
       prec
 
-instance : ToString ONote :=
-  ⟨toString⟩
-
-instance : Repr ONote where
-  reprPrec o prec := repr' prec o
```
Comment (eric-wieser) [inline Mathlib/SetTheory/Ordinal/Notation.lean:94]: I'd argue this style is better

File: Mathlib/SetTheory/Ordinal/Notation.lean [hunk 2]
```diff
@@ -100,42 +98,34 @@ instance : Preorder ONote where
...
-
--- Porting note (#11467): during the port we marked these lemmas with `@[eqns]`
--- to emulate the old Lean 3 behaviour.
-
-@[simp] theorem ofNat_zero : ofNat 0 = 0 :=
-  rfl
+instance : NatCast ONote :=
+  ⟨fun n ↦ match n with
+    | 0 => 0
+    | Nat.succ n => oadd 0 n.succPNat 0⟩
```
Comment (eric-wieser) [inline Mathlib/SetTheory/Ordinal/Notation.lean:110]:
```suggestion
instance : NatCast ONote where
  natCast
  | 0 => 0
  | Nat.succ n => oadd 0 n.succPNat 0
```

File: Mathlib/SetTheory/Ordinal/Notation.lean [hunk 3]
```diff
@@ -100,42 +98,34 @@ instance : Preorder ONote where
...
 
-@[simp] theorem ofNat_succ (n) : ofNat (Nat.succ n) = oadd 0 n.succPNat 0 :=
-  rfl
+/-- Convert a natural number to an ordinal notation -/
+@[deprecated (since := "2024-10-17")]
+def ofNat (n : ℕ) : ONote := n
 
-instance nat (n : ℕ) : OfNat ONote n where
-  ofNat := ofNat n
```
Comment (eric-wieser) [inline Mathlib/SetTheory/Ordinal/Notation.lean:128]: Where did this go?

File: Mathlib/SetTheory/Ordinal/Notation.lean [hunk 4]
```diff
@@ -100,42 +100,37 @@ instance : Preorder ONote where
...
-@[simp] theorem ofNat_succ (n) : ofNat (Nat.succ n) = oadd 0 n.succPNat 0 :=
-  rfl
+@[simp] theorem natCast_zero : (0 : ℕ) = (0 : ONote) := rfl
+@[simp] theorem natCast_succ (n : ℕ) : n.succ = oadd 0 n.succPNat 0 := rfl
+theorem natCast_one : (1 : ℕ) = (1 : ONote) := rfl
 
-instance nat (n : ℕ) : OfNat ONote n where
-  ofNat := ofNat n
+instance (n : ℕ) : OfNat ONote n where
+  ofNat := n
```
Comment (eric-wieser) [inline Mathlib/SetTheory/Ordinal/Notation.lean:123]: For these instances you should at least have a lemma `natCast_ofNat : natCast (noindex (OfNat.ofNat n)) = OfNat.ofNat n`

(assuming you don't get this for free with a later `AddCommMonoidWithOne` instance).

File: Mathlib/SetTheory/Ordinal/Notation.lean [hunk 5]
```diff
@@ -119,18 +119,25 @@ def ofNat (n : ℕ) : ONote := n
...
 theorem repr_zero : repr 0 = 0 := repr_natCast 0
+theorem repr_one : repr 1 = 1 := mod_cast repr_natCast 1
 
-@[deprecated (since := "2024-10-17")] alias repr_ofNat := repr_natCast
+instance (n : ℕ) : OfNat ONote n where
+  ofNat := n
+
+@[simp]
+theorem natCast_ofNat (n : ℕ) [n.AtLeastTwo] :
+    (no_index (OfNat.ofNat n) : ONote) = OfNat.ofNat n :=
```
Comment (eric-wieser) [inline Mathlib/SetTheory/Ordinal/Notation.lean:135]: This lemma is a tautology

Comment (eric-wieser) [reply Mathlib/SetTheory/Ordinal/Notation.lean:135]: (it also doesn't mention a cast; did you mean for one side to be case to `Ordinal`?

Comment (eric-wieser) [reply Mathlib/SetTheory/Ordinal/Notation.lean:135]: What type do you intend the RHS to be?

Comment (eric-wieser) [reply Mathlib/SetTheory/Ordinal/Notation.lean:135]: Maybe you meant
```suggestion
    ↑(no_index (OfNat.ofNat n) : ℕ) = (OfNat.ofNat n : ONote) :=
```

**Comment by vihdzp** [comment]
closing as I plan to completely overhaul this file in the near future

**Pull request closed without merging**